### PR TITLE
[SPARK-17017][Follow-up][ML] Refactor of ChiSqSelector and add ML Python API.

### PR DESCRIPTION
**Pull request by yanboliang**
## What changes were proposed in this pull request?
#14597 modified `ChiSqSelector` to support `fpr` type selector, however, it left some issue need to be addressed:
- We should allow users to set selector type explicitly rather than switching them by using different setting function, since the setting order will involves some unexpected issue. For example, if users both set `numTopFeatures` and `percentile`, it will train `kbest` or `percentile` model based on the order of setting (the latter setting one will be trained). This make users confused, and we should allow users to set selector type explicitly. We handle similar issues at other place of ML code base such as `GeneralizedLinearRegression` and `LogisticRegression`.
- Meanwhile, if there are more than one parameter except `alpha` can be set for `fpr` model, we can not handle it elegantly in the existing framework. And similar issues for `kbest` and `percentile` model. Setting selector type explicitly can solve this issue also.
- If setting selector type explicitly by users is allowed, we should handle param interaction such as if users set `selectorType = percentile` and `alpha = 0.1`, we should notify users the parameter `alpha` will take no effect. We should handle complex parameter interaction checks at `transformSchema`. (FYI #11620)
- We should use lower case of the selector type names to follow MLlib convention.
- Add ML Python API.
## How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by srowen** [comment]
Oh I see. I trust your judgment on this, just wish we could have gotten your review on the original PR. @mpjlu what do you think?

**Comment by mpjlu** [comment]
Hi @srowen and @yanboliang ; Thanks for your following up PR. 
I partly agree with your comments on 17017. 
**1. "if users both set numTopFeatures and percentile, it will train kbest or percentile model based on the order of setting (the latter setting one will be trained). This make users confused, and we should allow users to set selector type explicitly."**
For the user confused you mentioned here, I think the main reason is function name.  I have changed the function name of setAlpha to setFPR in SPARK-17645.  setNumTopFeature should be setKBest.
By this change, it can be much clear. 
For example, setKBest(100), setPercentile(0.1), setFPR(0.05). The selection type and parameters is very clear by one function.  
But for your method, user have to strike "setSelectorType("KBest").setNumTopFeatures(100)" to do the same thing as "setKBest(100)"
**2. "if there are more than one parameter except alpha can be set for fpr model, we can not handle it elegantly in the existing framework. And similar issues for kbest and percentile model. "**
I cannot think out any other parameters for fpr, kbest, percentile now. But if there is, I think it is just the same thing as your method. for example, setKBest(100).setOther(),, 
I agree with you for other change. Thanks very much.

File: mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala [hunk 1]
```diff
@@ -143,13 +149,13 @@ final class ChiSqSelector @Since("1.6.0") (@Since("1.6.0") override val uid: Str
         case Row(label: Double, features: Vector) =>
           OldLabeledPoint(label, OldVectors.fromML(features))
       }
-    var selector = new feature.ChiSqSelector()
-    ChiSqSelectorType.withName($(selectorType)) match {
-      case ChiSqSelectorType.KBest =>
+    val selector = new feature.ChiSqSelector()
+    $(selectorType) match {
+      case OldChiSqSelector.KBest =>
         selector.setNumTopFeatures($(numTopFeatures))
```
Comment (mpjlu) [inline mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala]: Do you need to set SelectorType here?

File: mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala [hunk 2]
```diff
@@ -143,13 +149,13 @@ final class ChiSqSelector @Since("1.6.0") (@Since("1.6.0") override val uid: Str
...
-      case ChiSqSelectorType.Percentile =>
+      case OldChiSqSelector.Percentile =>
         selector.setPercentile($(percentile))
-      case ChiSqSelectorType.FPR =>
+      case OldChiSqSelector.FPR =>
         selector.setAlpha($(alpha))
```
Comment (mpjlu) [inline mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala]: ditto

File: mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala [hunk 3]
```diff
@@ -160,6 +166,12 @@ final class ChiSqSelector @Since("1.6.0") (@Since("1.6.0") override val uid: Str
 
   @Since("1.6.0")
   override def transformSchema(schema: StructType): StructType = {
+    val otherPairs = OldChiSqSelector.supportedTypeAndParamPairs.filter(_._1 == $(selectorType))
```
Comment (mpjlu) [inline mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala]: == or != ?

File: mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala (hunk 1 above)
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala]: Yes, updated.

File: mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala (hunk 3 above)
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala]: Sorry, typo.

File: mllib/src/test/scala/org/apache/spark/mllib/feature/ChiSqSelectorSuite.scala [hunk 4]
```diff
@@ -76,7 +76,7 @@ class ChiSqSelectorSuite extends SparkFunSuite with MLlibTestSparkContext {
         LabeledPoint(1.0, Vectors.dense(Array(4.0))),
         LabeledPoint(1.0, Vectors.dense(Array(4.0))),
         LabeledPoint(2.0, Vectors.dense(Array(9.0))))
-    val model = new ChiSqSelector().setAlpha(0.1).fit(labeledDiscreteData)
+    val model = new ChiSqSelector().setSelectorType("fpr").setAlpha(0.1).fit(labeledDiscreteData)
```
Comment (yanboliang) [inline mllib/src/test/scala/org/apache/spark/mllib/feature/ChiSqSelectorSuite.scala]: Added ML test case.

**Comment by yanboliang** [comment]
@mpjlu The most important cause of this change is that the fit/train model should not dependent on the order of users setting params. In other words, users should get the same model whether set A following B or B following A. Thanks!

**Comment by mpjlu** [comment]
Hi @yanboliang , got it. Thanks.

**Comment by srowen** [comment]
I'm OK with it. @mpjlu sounds like you approve?

**Comment by mpjlu** [comment]
hi @srowen .
My understand of yanbo's comments here is,
if user use chSqSelector like this:
model1 = new ChiSqSelector().setFPR(0.05).setKBest(100).fit(data)
model2 = new ChiSqSelector().setKBest(100).setFPR(0.05).fit(data)
model1 will be different with model2. so the model is dependent on the order of users setting params.
Actually, user should not use ChiSqSelector like this. One just need to set one SelectorType/Parameter is ok.  But if one don't know ChiSqSelector, he may do like this. So yanbo think this is a problem. 

In this PR, setFPR(0.05) is split to two functions: setSelectorType("fpr").setAlpha(0.05). This maybe clear to the user.  
By the principle of software development: one function do one thing, I am ok with this change.  
But from user experience,  I like the spark-17017 method.

**Comment by srowen** [comment]
OK, I could also support either behavior. After all, for any component, `.setFoo(x).setFoo(y)` also creates a different model if the order is swapped, so I am not so clear that's a 'problem'.

**Comment by yanboliang** [comment]
@srowen @mpjlu 
Another important reason for this change: it's error prone for Python ML API. 
```
def __init__(self, numTopFeatures=50, featuresCol="features", outputCol=None,
                   labelCol="label", selectorType="kbest", percentile=0.1, alpha=0.05):
......
def setParams(self, numTopFeatures=50, featuresCol="features", outputCol=None,
                         labelCol="labels", selectorType="kbest", percentile=0.1, alpha=0.05):
......
```
If users are not very familiar with ```ChiSqSelector```, they are likely to set all parameters following the API docs. The output model is also relevant with the arguments order. Users not aware of the order of arguments for the Python API is a very strong possibility. Thanks.

**Comment by yanboliang** [comment]
And you can also refer all other Estimator in ML, even you swap the arguments setting order, you still get the same model. Thanks.

**Comment by mpjlu** [comment]
Thanks, this looks good to me.

**Comment by mpjlu** [comment]
Hi @srowen , sorry for forgetting update the doc and python/ml/feature.py in last PR.
This pr has added ml/feature.py. It looks good to me.
Thanks

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**